### PR TITLE
Integrate suppliers form in product options

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Moved code from another pr (#22703) for better readability and will continue here. I'll add more info when done.
| Type?             | refacto
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | part of #19683
| How to test?      | to be defined
| Possible impacts? | 

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/22743)

### CODE REVIEW

**Comment by zuk3975** [comment]
blocked by #22703

**Comment by matks** [comment]
A nice suggestion from @jolelievre : if user has a product with 4 suppliers and 1000 combinations, this will display 4000 fields 😄  which can prevent the correct loading of the page.

So quick and cheap trick to avoid this: we should NOT load this data if we expect this will crash. So we need to cut the fetching data logic when we think we'll go higher than what `post_max_size` AND/OR `max_input_vars` can handle.

So we need to find out what "average" or "standard `post_max_size` and `max_input_vars` values are and put a threshold to make sure it is never reached.

File: admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js]:
```suggestion
```

File: admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js:25]:
```suggestion

import ProductMap from '@pages/product/product-map';

```

**Comment by zuk3975** [comment]
todo (some of these might go to another PR):
* Supplier list is not ordered (inputs are appended one after another without caring about carrierIds selection order). Instead of appending/removing items one by one, we should consider having some centralized storage for object and rerender the list all the time (like in standard mvvm frameworks), although this requires lots of time and effort)
* All combinations are loaded on page load (performance stuff)
* I have modified CombinationRepository methods a little bit (removed redundant array level), so **GetCombinationsForEditing** might need little adjustments (thats why ProductCombination **behats are failing**)
* Saving doesnt work - duplicate entry. (depends too much on javascript - ProductSuppliersCommandBuilder needs some attention).
* There are some "inception" loops that might be optimized (in ProductFormDataProvider, GetProductSuppliersForEditing| GetProductSupplierOptions)
* :hankey: code quality - needs refactoring (e.g. manager.js should be split into renderer.js as in orders page)
* Styling might be adjusted as well
* Needs tweaking to work in **ADD NEW PRODUCT** page as well (as it requires product name and id now - this means each supplier form PRODUCT_NAME should be dynamic)

:warning:  :hankey: Overall clearly its something wrong with this feature, its absolute mess in every aspect including code quality, performance and UX. some reasons:
* **involves combinations** (requires them all actually for each supplier), so with just a couple combinations its unreadable and almost unusable
* has no pagination/filters/sorting etc. (and I have no idea how would it look like. (I suppose this would be benefitial only for combinations)?
* forces merchant to have a supplier for each combination (even if fields are empty, empty productSuppliers will be created anyway)
* even tho product has combinations, there is still ProductSupplier without combination created, but it isn't reflected in view and cannot be modified.

**Comment by MatShir** [comment]
Here, what's have been decided from the last meeting.
If it is a simple product, virtual product, or a pack
Then we keep it as is. The supplier is handle in the options tab with the checkboxes to select the suppliers and the option selector to select the default supplier along with the prices' lists by suppliers.

If it is a combination product, 
Then we only keep in the options tab with the checkboxes to select the suppliers and the option selector to select the default supplier.
And we display on each combination the list of the prices' lists by suppliers. 

@zuk3975 I'm seeing Tristan shortly to discuss the design. Should I make an issue about this new behavior? 
The specs will be updated today FYI.

**Comment by zuk3975** [comment]
> Should I make an issue about this new behavior?

Thanks, @MatShir , yes I think we should have an issue for it :thinking:

**Comment by zuk3975** [comment]
there is still one bug I don't know how to resolve.
After submit, if there are any errors in suppliers form, then supplier references form collection data dissapears and it results showing "__SUPPLIER_NAME__" placeholder  instead of actual supplier name in table :/


**EDIT**
idk how or why, but the bug seems to be gone again :man_shrugging:

File: admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js
Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js]: Indeed the import must be after the header, I have similar issues with my IDE ^^

File: admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js]:
```suggestion
  addSupplier(supplier) {
```

File: admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js]:
```suggestion
  removeSupplier(supplierId) {
```

File: admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js]:
```suggestion
    this.defaultSupplierData = this.collectDefaultSupplierData();
```

Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js]: I find this naming clearer, but it's just my opinion and debatable

File: admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js]: Although I understand the purpose I wonder if naming can be improved, maybe `saveSuppliersData` `backupSuppliersData` or something similar which would indicate the purpose of this action To save data for future rendering

File: admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js]: You should add a comment here to explain that the suppliers' data is kept in memory so that it can be used to refresh/render them later

File: admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js]:
```suggestion
        supplierId: supplier.id,
```
This one doesn't need to be fetched from DOM

File: admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js
Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js]: Also add a comment in the function indicating that we use a `deleted` state boolean to filter its rendering, but it allows to keep previous data in memory in case it is added again

File: admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js]: Usually in a collection of elements the index used is not related to the an ID, this avoids matching errors which is why I suggested to add the `supplier_id` as a real field in the sub form type This prevents any confusion
Therefore the placer holder should rather be `__SUPPLIER_INDEX__` and be replaced by 0, 1, 2, ...

Although in your case it would work because in the `ProductFormDataProvider` you correctly used the ID as the key so everything is consistent along the workflow and it works as expected But I'm always wary of what might happen when you have not consecutive IDs used as key, which result in your request containing `product[supplier][1]` and `product[supplier][5]` for example

It feels cleaner when data are formatted as `product[supplier][0]` `product[supplier][1]`

Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js]: When you look at the Symfony doc they advise using an incrementing index https://symfony.com/doc/3.4/form/form_collections.html

File: admin-dev/themes/new-theme/js/pages/product/product-map.js
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/product-map.js]: Wow you really took this mapping to another level 😅

I fear we are going to have a very huge  object which makes me think about something we tried to do a bit too late in the Order page mapping. Split the selectors by domain in a hierarchical object instead of all of them at the same level. Something like this:

```js
export default {
  productForm: 'form[name=product]',
  supplier: {
    block: '#productSuppliers',
    referencesBlock: '#product_suppliers_supplier_references',
  }
}
```

You can then use `ProductMap.supplier.block` This will make more easy to read, it will be less a mess

File: src/Adapter/Form/ChoiceProvider/SupplierNameByIdChoiceProvider.php
Comment (jolelievre) [inline src/Adapter/Form/ChoiceProvider/SupplierNameByIdChoiceProvider.php:34]: Can you add unit tests for this class? :trollface:
(just kidding of course)

File: src/Adapter/Product/QueryHandler/GetProductSupplierOptionsHandler.php
Comment (jolelievre) [inline src/Adapter/Product/QueryHandler/GetProductSupplierOptionsHandler.php]: I thought we said handlers were not supposed to use other handlers internally any more
I guess this one was made after we made that decision, it would deserve a little refacto I guess (in another PR)

File: src/Adapter/Product/QueryHandler/GetProductSuppliersForEditingHandler.php
Comment (jolelievre) [inline src/Adapter/Product/QueryHandler/GetProductSuppliersForEditingHandler.php]: Not sure I understand the purpose here..

File: src/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProvider.php
Comment (jolelievre) [inline src/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProvider.php]: Since there is only one reference used by the form maybe we should loop through the whole list but only use the first element no? Maybe it would even be worth modifying the query handler result or have two commands one Dedicated to Simple product and another for combinations product?

But anyway there's a problem with this loop it actually replaces the `$suppliersData['supplier_references'][$supplierId]['product_supplier']` part in each iteration which actually results in returning the data from last combination

File: src/PrestaShopBundle/Form/Admin/Sell/Product/ProductType.php
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/ProductType.php]: You shouldn't need this, it was not necessary on all other form types Since the template is supposed to call `form_widget(suppliersForm)` the label is not displayed

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig]: I think now that the collection is more simple this `table` initialization part can be moved into the `_product_suppliers_supplier_references_row` You shouldn't be obliged to handle this manually in this template, by relying on Symfony customization feature you can simplify this template

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig:32]: You shouldn't handle the display field by field, just using `form_widget(suppliersForm)` should be enough, or the system we want to introduce to override/extend forms won't be appliable any more

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig]: The id here shouldn't be required since you already have an ID on the input itself

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig
Comment (jolelievre) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig]: If you need help to simplify this form code I can give you a hand

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/tabs.html.twig
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/tabs.html.twig]:
```suggestion
{% set isOptionsTabValid = productForm.options.vars.valid and productForm.suppliers.vars.valid %}
```

It's simpler this way

**Review by jolelievre — COMMENTED** [review]
Please don't hate me 😅

**Comment by jolelievre** [comment]
Also seems like `TypedRegexValidatorTest` is failing for some reason

File: admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js
Comment (zuk3975) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js]: I would agree, but `default` might be confused with default supplier selection, that is why I use `new` instead.

Comment (zuk3975) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js]: changed to `defaultDataForSupplier`

File: admin-dev/themes/new-theme/js/pages/product/product-map.js
Comment (zuk3975) [reply admin-dev/themes/new-theme/js/pages/product/product-map.js]: why couldn't we simply use another file, smth like suppliers-map instead?

File: src/Adapter/Product/QueryHandler/GetProductSuppliersForEditingHandler.php
Comment (zuk3975) [reply src/Adapter/Product/QueryHandler/GetProductSuppliersForEditingHandler.php]: you can get suppliers for standard product or for combinations. 
If you get it for combinations by default you will still see one additional productSupplier which is associated with product id without combination :sweat: 
idk why its done this way, but Im afraid to dig deeper, so I just have to filter that one out here.

File: src/Adapter/Product/QueryHandler/GetProductSupplierOptionsHandler.php
Comment (zuk3975) [reply src/Adapter/Product/QueryHandler/GetProductSupplierOptionsHandler.php]: yeah Im sure we did replace those usages in commands, but not queries.
btw this query might need a refacto after the use case have changed :thinking: 
I wonder if it deserves separate pr?

File: admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js
Comment (zuk3975) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js]: Yeah at first I tried using indexes, but I got stuck somewhere. Matching everything by id seemed way easier and required less code, I've checked that everything works like this and stayed to it. You say its worth refactoring to use indexes?

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig
Comment (zuk3975) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig]: I needed to wrap them up into a block to get all inputs, this was the easiest way :smile:

File: src/Core/ConstraintValidator/Constraints/TypedRegex.php
Comment (Progi1984) [inline src/Core/ConstraintValidator/Constraints/TypedRegex.php]: Please define the scope of the constant.

**Review by Progi1984 — COMMENTED** [review]
Very good job. :tada: 

Please add unit test for `src/Core/Domain/Product/Supplier/Query/GetProductSuppliersForEditing.php`, `src/Core/Form/IdentifiableObject/CommandBuilder/Product/ProductSuppliersCommandBuilder.php` and `src/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProvider.php`

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig
Comment (Julievrz) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig]:
```suggestion
          <th>{{ 'Supplier'|trans({}, 'Admin.Global') }}</th>
```

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig
Comment (zuk3975) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig]: okay, I've resolved most of the "unarguable" feedback, so as discussed I can leave some of these js and templating stuff to you, thanks! (this Pr will benefit from fresh eyes :stuck_out_tongue_closed_eyes: )

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/form_fields.html.twig
Comment (zuk3975) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/form_fields.html.twig]: btw @jolelievre we could rename this file to something specific for suppliers and maybe mention `theme` in name?
product_suppliers_theme? :man_shrugging:

Comment (jolelievre) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/form_fields.html.twig]: I named it `form_fields` originally because I thought it could contain other form overrides for other parts But I guess you're right it's probably cleaner to have a dedicated file for each theme

File: tests/Unit/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProviderTest.php
Comment (jolelievre) [inline tests/Unit/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProviderTest.php:111]: @zuk3975 you nailed this part! I love how you improved the provider functions I really have some progress to make regarding the use of `yield` ^^

I tried doing something similar some time ago (mixing loop and yield) but without success, it's quite interesting to see it in action!

File: src/Adapter/Product/Repository/ProductSupplierRepository.php
Comment (PierreRambaud) [inline src/Adapter/Product/Repository/ProductSupplierRepository.php:51]:
```suggestion
    protected $connection;
```

File: src/Adapter/Product/Repository/ProductSupplierRepository.php
Comment (PierreRambaud) [inline src/Adapter/Product/Repository/ProductSupplierRepository.php:56]:
```suggestion
    protected $dbPrefix;
```

File: src/Adapter/Product/Repository/ProductSupplierRepository.php
Comment (jolelievre) [reply src/Adapter/Product/Repository/ProductSupplierRepository.php:51]: This is a nice idea, I think there is the same issue in other repositories so I'll make a general PR for all of them once this one is merged 😉

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig
Comment (NeOMakinG) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Blocks/suppliers.html.twig:32]: I agree, overwise, JS looks fine to me

File: admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js
Comment (matks) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js:47]: I like this function naming

File: admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js
Comment (matks) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-suppliers-manager.js:204]: We'll need to document when we hit feature freeze 📚

File: src/Adapter/Form/ChoiceProvider/SupplierNameByIdChoiceProvider.php
Comment (matks) [reply src/Adapter/Form/ChoiceProvider/SupplierNameByIdChoiceProvider.php:34]: And attach the code coverage? 😏

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Form/suppliers_form_theme.html.twig
Comment (matks) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Form/suppliers_form_theme.html.twig]:
```suggestion
  {# We put prototype name in a data attribute, so the JS code can get it dynamically #}
```

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Form/suppliers_form_theme.html.twig
Comment (matks) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Form/suppliers_form_theme.html.twig:25]: Hoho, so we're providing a custom form theme block implementation for product page? 😄 

I like it, but we'll need to be extra explicit about it 😄  not the easiest thing to understand

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Form/suppliers_form_theme.html.twig
Comment (matks) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Form/suppliers_form_theme.html.twig]:
```suggestion
          {# We add column for children that are not hidden type, unless their label is not empty (like supplier_name) #}
```

**Review by matks — COMMENTED** [review]
Nothing to say more than my colleagues. Just two little writing mistakes. Great job!

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Form/suppliers_form_theme.html.twig
Comment (jolelievre) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Form/suppliers_form_theme.html.twig:25]: This one is not for the whole product page, it's only assigned to the suppliers sub form This is a classic way to use Symfony twig override system, nothing to document further I'd say 🤔

**Review by matks — APPROVED** [review]

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Form/suppliers_form_theme.html.twig
Comment (matks) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Product/Form/suppliers_form_theme.html.twig:25]: Even if we do know Symfony and Twig well, not every PrestaShop contributors and users do 😉  .

No need to duplicate Symfony documentation, just need to say "and this `suppliers_form_theme.html.twig` file is actually an extension of PrestaShop default twig form theme, you can learn more about it from [this link]" :wink:

**Review by PierreRambaud — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Thank you @zuk3975

**Pull request merged**
